### ISSUE TITLE
1) enable reporter for imported contracts 2) gas-reporter discrepancies with mainnet

### ISSUE THREAD

**Issue by gitpusha** [cgewecke/hardhat-gas-reporter#46]
Hi @cgewecke ,

I realized that `hardhat-gas-reporter` only seems to report for contracts who were compiled with `hardhat` and thus have certain files associated with them in `artifacts/` .

Is this correct? 

In our integration testing we make use of `forking` and thus instantiate many contracts with `await ethers.getContractAt(abi, addresss)` , where the `abi` is in a `pre-compiled/` folder.

It would be amazing if `hardhat-gas-reporter` would also report gas usage of methods called on those instances. 

What would need to change for that to happen?

**Comment by cgewecke** [comment]
Hi @gitpusha,

Yes, agree this would be really useful and should be straightforward to implement.  

If there was a gasReporter option like `remoteContracts` which took an array of
```ts
[
  {
    contractName: <string>
    address: <string>
    abi: <string[]>
  },
]
```
would you be able to supply that info for the on-chain contracts before the tests run?

Do you have a repo I could test an option like this with? 


[1]: https://github.com/cgewecke/hardhat-gas-reporter/blob/5ec5a7661bf6087459279ac1b5033d35e5727305/src/index.ts#L30-L61
[2]: https://github.com/cgewecke/hardhat-gas-reporter/blob/5ec5a7661bf6087459279ac1b5033d35e5727305/src/index.ts#L84

**Comment by gitpusha** [comment]
Yes, such an option would be great ! 

Here is a [repo](https://github.com/gelatodigital/gelato-instadapp) you can test this with. 

We have some massive integration tests in there. 

Maybe just test with [this](https://github.com/gelatodigital/gelato-instadapp/blob/master/test/0_setup-DSA-Gelato.test.js) one.

[Here](https://github.com/gelatodigital/gelato-instadapp/blob/40de5ce04c26832773127674f9e371c798b50b7a/test/0_setup-DSA-Gelato.test.js#L13) you can find all the pre-compiles this test uses.

And [here](https://github.com/gelatodigital/gelato-instadapp/blob/40de5ce04c26832773127674f9e371c798b50b7a/hardhat.config.js#L64) are the addresses of the pre-compiles. 

If you wanna, you can release and `rc` and I can test this stuff too.

**Comment by cgewecke** [comment]
Have published an experimental `rc` version of this:
```
yarn install hardhat-gas-reporter@rc --dev
```

To test, I added the following to your `hardhat.config.js`
```js
const fs = require('fs');
const path = require('path');

// Load remote contract address and ABI data  
const remoteContracts = [];
const preCompilesPath = 'pre-compiles';
for (const fileName of fs.readdirSync(preCompilesPath)){
  const artifactPath = path.join(process.cwd(), preCompilesPath, fileName)
  const preDeployed = require(artifactPath);
  const address = mainnetDeployments[preDeployed.contractName]
  if (address){
    remoteContracts.push({
      name: preDeployed.contractName,
      abi: preDeployed.abi,
      address: address
    })
  }
}
```
and added `remoteContracts` to the gasReporter options
```diff
 gasReporter: {
     enabled: process.env.REPORT_GAS ? true : false,
     maxMethodDiff: 25,
     coinmarketcap: process.env.COINMARKETCAP_API_KEY,
+    remoteContracts
  },
```

However I'm not sure the results are what is expected. Only caught a couple of additional method calls, to ConnectAuth and InstaConnectors. 

Are you expecting more data here? Do you have lots of calls routed through a proxy or call-forwarding contract? 


![Screen Shot 2020-11-16 at 5 57 33 PM](https://user-images.githubusercontent.com/7332026/99336652-47083f80-2836-11eb-81db-0df9f05e519c.png)

**Comment by gitpusha** [comment]
Amazing! Thanks so much for the hands-on support.

The table looks pretty good to me. I was opening this issue with generality in mind but I only had a couple of methods missing. So yes, looks good to me. 

 > Do you have lots of calls routed through a proxy or call-forwarding contract?

Yes, that is why there aren't too many new methods shown, but that's expected. 

I will use your rc and test it myself. When do you think will you make it a stable release? 

I see that your method only requires:

```javascript
    remoteContracts.push({
      name: preDeployed.contractName,
      abi: preDeployed.abi,
      address: address
```

Hence a `name` and `abi` first-level property on whatever artifact is in my `pre-compiles` folder. This basically means that as long as my pre-compiles have those 2 properties it should work right? They don't need to be a hardhat style artifact, correct?

**Comment by cgewecke** [comment]
> Yes, that is why there aren't too many new methods shown, but that's expected.

There's an [option available][4] which lets you to pass the reporter a method to resolve the real identity of proxy calls. It's a little complicated (but it's theoretically possible...). 

> This basically means that as long as my pre-compiles have those 2 properties it should work right?

Yes, exactly. 

> when we measured our gas used manually with ethers.js and the gasUsed property on a TransactionReceipt ...

Oh! Super interesting. How did you measure manually? 

The hardhat reporter really just wraps the Ethers provider - it would be great to track down exactly where the gap you're seeing comes from. We run the eth-gas-reporter test suite against ganache and geth and their results differ by a few hundred gas for very simple methods.

+ [geth][1]    
+ [ganache][2]

Also, recently [a bug was reported][3] at ganache that gas readings are substantially off when running in forked mode. 

(Don't know how closely HH tracks geth - should add the same test to this repo.)


[1]: https://travis-ci.org/github/cgewecke/eth-gas-reporter/jobs/744087234#L835
[2]: https://travis-ci.org/github/cgewecke/eth-gas-reporter/jobs/744087232#L786
[3]: https://github.com/trufflesuite/ganache-core/issues/625
[4]: https://github.com/cgewecke/eth-gas-reporter/blob/master/docs/advanced.md#resolving-method-identities-when-using-proxy-contracts

**Comment by cgewecke** [comment]
Missed a question...
>  When do you think will you make it a stable release?

I'm not sure...it might be a couple weeks. There are a couple other fixes I'd like to get in here

**Comment by gitpusha** [comment]
> There's an option available which lets you to pass the reporter a method to resolve the real identity of proxy calls. It's a little complicated (but it's theoretically possible...).

I had a look at it. Pretty cool. But actually we are not making typical Proxy calls via a fallback method. The problem for gas reporting is rather that we use a single entry point function for different payloads. This function wraps around many different paths of execution. And we want to measure gas for those different execution paths.

So even though the important stuff for gas-reporting is hidden in the payload, the gas-reporter tool will average out gasUsed across all our tested execution paths, which renders it kind of useless in that context because we want to differentiate between them distinctly. 

The difference in execution paths are nested deep down, so I cannot think of a simple way to write a general Resolver for this. I think it's easier for now to just Mock the different paths by giving them all a distinct method name and then submit the respective payloads via this distinct method name, thereby tricking the tool into reporting gas for these different paths. The real system will only have one entry method though, not several. 

> Oh! Super interesting. How did you measure manually?

Just using ethers TransactionReceipt.gasUsed property. Probably similar to what you do. 

> Also, recently a bug was reported at ganache that gas readings are substantially off when running in forked mode.

We also run our tests in `hardhat` forkmode. So maybe something similar is going on here. The measurements were off by a significant margin (something like 600k gas) . 

I will probably have some files where you can review this soon and share them with you.

**Comment by cgewecke** [comment]
> I think it's easier for now to just Mock the different paths by giving them all a distinct method name and then submit the respective payloads via this distinct method name, thereby tricking the tool into reporting gas for these different paths.

Yes this makes a lot of sense.

> I will probably have some files where you can review this soon and share them with you.

Ok cool, yes I'd really like to see this and get reproduction steps for the difference.

**Comment by gitpusha** [comment]
> Ok cool, yes I'd really like to see this and get reproduction steps for the difference.

Ok, so I just tried to reproduce myself and for me the manual ethers method 

```javascript
const txResponse = await contract.method();
const { gasUsed } = await txResponse.wait();
console.log(gasUsed);
```

yields exactly the same results as `hardhat-gas-reporter` which should be expected and which is a relief of sorts. My colleague reported the disrepancy to me but I wasnt able to reproduce myself. 

However, what is clear is that both `ethers` and `hardhat-gas-reporter` report wrong values apparently because on mainnet the same TX consumes much more gas. 

And I believe that my colleague really did find these disrepancies in testing because we went with the higher gas values he found during testing using the manual `gasUsed` method and they turn out to be close to mainnet gasUsed. Luckily we hardcoded those ones into our contract before we deployed. 

However, I am still confused why now both the tool and the manual method report the same apparently too-low gasUsed value. 

We run this stuff on a hardhat-network fork of mainnet.

**Comment by gitpusha** [comment]
Actually new finding:

I added some `hardhat` `console.log` to the solidity file to see what that has to say about `gasUsed` 

![image](https://user-images.githubusercontent.com/36712489/99689983-e708d780-2a87-11eb-9b8f-77ec79bd1cc8.png)

![image](https://user-images.githubusercontent.com/36712489/99690005-ec662200-2a87-11eb-80ab-be75281c148a.png)

Clearly hardhat network console.logs also think there is more gas consumed than both ethers.js and hardhat-gas-reporter. 

Is there ever a scenario where hardhat-gas-reporter's values could diverge from vanilla ethers.js values? No right? They are the same values read from the same thing basically right?

**Comment by cgewecke** [comment]
Thanks for this!

> Clearly hardhat network console.logs also think there is more gas consumed than both ethers.js and hardhat-gas-reporter.

Could you give a quick example of where/what/how you're console.logging in Solidity? (It's possible the logging itself distorts gas values a little at runtime.)

Also, is there a difference in the optimization levels you use to deploy to mainnet vs. local testing?

> Is there ever a scenario where hardhat-gas-reporter's values could diverge from vanilla ethers.js values? No right? They are the same values read from the same thing basically right?

Yes they are supposed to be, there could be a bug somewhere though.

**Comment by gitpusha** [comment]
[Here](https://github.com/gelatodigital/gelato-instadapp/blob/92f8b6983ab316eee72f7eef9439e108d71f1f7f/contracts/__mocks__/gelato/MockDebtBridgeETHBExecutor.sol#L53) in the contract I shared you can see the `console.logs`. 

```solidity
    function execViaRoute0AndOpenVault(TaskReceipt memory _taskReceipt) public {
        uint256 gasLeft = gasleft();
        IGelatoCore(gelatoCore).exec(_taskReceipt);
        console.log(
            "Gas Cost execViaRoute0AndOpenVault: %s",
            gasLeft - gasleft()
        );
    }
```

**Comment by gitpusha** [comment]
This got better with the new hardhat release: https://github.com/nomiclabs/hardhat/releases/tag/hardhat-core-v2.0.4

https://github.com/nomiclabs/hardhat/issues/990


The tool now reports this:

![image](https://user-images.githubusercontent.com/36712489/100514409-be15d000-3174-11eb-8420-3de6a5c5ce18.png)

Instead of the [previous lesser values](https://github.com/cgewecke/hardhat-gas-reporter/issues/46#issuecomment-730468335). 

But

**Comment by gitpusha** [comment]
Reopening because I still observe about 10% higher gasUsed values on mainnet when comparing with hardhat network in forking mode.

**Comment by cgewecke** [comment]
@gitpusha The `remoteContracts` option is now published with `1.0.2`. And I suspect the discrepancy in gas readings is only fixable at hardhat. 

Going to close this for now. Thanks for reporting all this stuff though.

**Comment by gitpusha** [comment]
Ok !
